File: typescript/src/verify.ts
/**
 * RFC 9421 + RFC 9530 verification top-level.
 *
 * Mirror of the Python algovoi_rfc9421_verifier.verify module.
 * Uses @noble/ed25519 for Ed25519 verification.
 */

import { webcrypto } from "node:crypto";

import * as ed25519 from "@noble/ed25519";

import {
  parseSignatureInput,
  parseSignatureValue,
  SignatureInputParseError,
} from "./parse.js";
import {
  buildSigningBase,
  SigningBaseError,
  type SigningBaseMode,
} from "./signing-base.js";
import {
  verifyContentDigest,
  ContentDigestError,
} from "./content-digest.js";
import { checkFreshness, FreshnessError } from "./freshness.js";
import { checkEd25519PublicKey, WeakKeyError } from "./keycheck.js";

// Node 18 compatibility. @noble/ed25519 v2's async hashing reaches for
// crypto.subtle, which Node 18 does not expose as a global (Node 20+ does).
// Without it verifyAsync throws "crypto.subtle must be defined", which the
// verify path below would otherwise swallow as a failed verification. Provide
// the WebCrypto global when it is absent so verification behaves identically
// across Node 18, 20, and 22.
if (typeof (globalThis as { crypto?: unknown }).crypto === "undefined") {
  (globalThis as { crypto?: unknown }).crypto = webcrypto;
}

export class VerifyError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "VerifyError";
  }
}

export type PublicKey = string | Uint8Array;

export interface VerifyResult {
  valid: boolean;
  signature_valid: boolean;
  content_digest_valid: boolean;
  signing_base: string;
  covered_components: string[];
  parameters: Record<string, string | number>;
  label: string;
  errors: string[];
}

export interface VerifyRequestInput {
  method: string;
  authority: string;
  path: string;
  headers: Record<string, string>;
  body: Uint8Array | Buffer | string;
  publicKey: PublicKey;
  scheme?: string;
  requireContentDigest?: boolean;
  requireAlgorithm?: string | null;
  /**
   * Signing-base mode. Default "rfc9421" (v0.3.0+) builds the full
   * RFC 9421 §2.5 signing base, including the mandatory
   * "@signature-params" line, so it verifies any RFC-compliant
   * signer (external fixtures, and the rfc9421_proxy_chain_v1
   * conformance set). This matches the Python verify_request default.
   * Set to "algovoi-v0" for the legacy v0.1.0 base (no
   * "@signature-params" line, @method lowercased) used by the
   * rfc9421_proxy_chain_v0 fixture.
   */
  mode?: SigningBaseMode;

  // --- Sprint A hardening (v0.4.0): freshness / replay / tag / alg ---
  /**
   * Current unix time in seconds. Injected so verification is deterministic in
   * tests and against static fixtures. Defaults to Math.floor(Date.now()/1000).
   */
  now?: number;
  /**
   * Reject when the signed `created` is older than this many seconds. Undefined
   * (the default) disables the age check, which is what keeps long-lived static
   * fixtures valid.
   */
  maxAgeSeconds?: number | null;
  /** Clock-skew tolerance for future-created and past-expires bounds (default 60). */
  maxSkewSeconds?: number;
  /** When true (default), a covered `expires` in the past is rejected. */
  enforceExpires?: boolean;
  /** When true, a signed `created` must be present even if maxAgeSeconds is off. */
  requireCreated?: boolean;
  /** When set, the signature's `tag` parameter must equal this value. */
  expectedTag?: string | null;
  /** When true, a `tag` parameter must be present. */
  requireTag?: boolean;
  /**
   * Allow-list of accepted signature algorithms (lowercased on compare).
   * Defaults to {"ed25519"}. An absent `alg` parameter is always rejected.
   */
  allowedAlgorithms?: Iterable<string>;
  /**
   * Single-use nonce store probe: (nonce, keyid) -> hasBeenSeen. Called only
   * after the signature verifies. A thrown error fails closed (never open).
   */
  nonceSeen?: (nonce: string, keyid: string) => boolean;
  /**
   * Required covered components (RC-1). When set, a signature whose covered set
   * omits any of these (e.g. "@method", "@authority", "@path") is rejected before
   * verification, closing the request-line-rewrite gap where a valid signature
   * over a narrow covered set leaves the rest of the request unsigned and mutable.
   * Case-insensitive. Omit to disable (no behaviour change).
   */
  requiredComponents?: string[];
}

function newResult(): VerifyResult {
  return {
    valid: false,
    signature_valid: false,
    content_digest_valid: false,
    signing_base: "",
    covered_components: [],
    parameters: {},
    label: "",
    errors: [],
  };
}

function fail(result: VerifyResult, msg: string): VerifyResult {
  result.errors.push(msg);
  result.valid = false;
  return result;
}

function publicKeyBytes(pk: PublicKey): Uint8Array {
  if (pk instanceof Uint8Array) {
    if (pk.length !== 32) {
      throw new VerifyError(
        `Ed25519 public key must be 32 bytes, got ${pk.length}`,
      );
    }
    return pk;
  }
  if (typeof pk === "string") {
    const hex = pk.startsWith("0x") ? pk.slice(2) : pk;
    if (!/^[0-9a-fA-F]+$/.test(hex)) {
      throw new VerifyError("public key hex is invalid");
    }
    if (hex.length !== 64) {
      throw new VerifyError(
        `Ed25519 public key hex must decode to 32 bytes (64 hex chars), got ${hex.length}`,
      );
    }
    const bytes = new Uint8Array(32);
    for (let i = 0; i < 32; i++) {
      bytes[i] = parseInt(hex.slice(i * 2, i * 2 + 2), 16);
    }
    return bytes;
  }
  throw new VerifyError("publicKey must be Uint8Array or hex string");
}

export async function verifySignature(
  signingBase: string,
  signatureBytes: Uint8Array,
  publicKey: PublicKey,
  algorithm: string = "ed25519",
): Promise<boolean> {
  if (algorithm.toLowerCase() !== "ed25519") {
    throw new VerifyError(`v0.1.0 supports ed25519 only; got ${algorithm}`);
  }
  if (signatureBytes.length !== 64) {
    throw new VerifyError(
      `Ed25519 signature must be 64 bytes, got ${signatureBytes.length}`,
    );
  }
  const pkBytes = publicKeyBytes(publicKey);
  // Trust-boundary key gate: reject non-canonical, off-curve, and small-order
  // public keys BEFORE verification. @noble/ed25519 (ZIP215) accepts small-order
  // keys, which enables signature-malleability / cross-key classes.
  try {
    checkEd25519PublicKey(pkBytes);
  } catch (e) {
    if (e instanceof WeakKeyError) throw new VerifyError(e.message);
    throw e;
  }
  const messageBytes = new TextEncoder().encode(signingBase);
  try {
    return await ed25519.verifyAsync(signatureBytes, messageBytes, pkBytes);
  } catch (e) {
    // A crypto-setup error (e.g. missing crypto.subtle) must not be silently
    // reported as an invalid signature. Surface it as VerifyError; only a
    // genuine verification mismatch returns false.
    if (
      e instanceof Error &&
      /crypto\.subtle|subtle must be defined/i.test(e.message)
    ) {
      throw new VerifyError(e.message);
    }
    return false;
  }
}

export async function verifyRequest(
  input: VerifyRequestInput,
): Promise<VerifyResult> {
  const result = newResult();

  const normHeaders: Record<string, string> = {};
  for (const [k, v] of Object.entries(input.headers)) {
    normHeaders[k.toLowerCase()] = v;
  }

  const siValue = normHeaders["signature-input"];
  if (!siValue) return fail(result, "Signature-Input header missing");
  const sValue = normHeaders["signature"];
  if (!sValue) return fail(result, "Signature header missing");

  let parsedSi;
  try {
    parsedSi = parseSignatureInput(siValue);
  } catch (e) {
    if (e instanceof SignatureInputParseError) {
      return fail(result, `Signature-Input parse error: ${e.message}`);
    }
    throw e;
  }
  result.label = parsedSi.label;
  result.covered_components = parsedSi.covered_components;
  result.parameters = parsedSi.parameters;

  // RC-1: caller policy on covered-component completeness. A signature over a
  // narrow covered set is cryptographically valid but may omit request-line
  // components (@method/@authority/@path); when the caller pins a required set, a
  // signature omitting any of them is rejected here rather than trusting a
  // narrowly-scoped signature over a rewritten request line. The covered list is
  // still returned either way. (Parity with the Python verifier.)
  if (input.requiredComponents && input.requiredComponents.length > 0) {
    const covered = new Set(
      parsedSi.covered_components.map((c) =>
        c.trim().replace(/^"+|"+$/g, "").toLowerCase(),
      ),
    );
    const missing = input.requiredComponents.filter(
      (c) => !covered.has(c.trim().toLowerCase()),
    );
    if (missing.length > 0) {
      const repr = missing
        .slice()
        .sort()
        .map((s) => `'${s}'`)
        .join(", ");
      return fail(result, `required covered components missing: [${repr}]`);
    }
  }

  let sigParsed;
  try {
    sigParsed = parseSignatureValue(sValue);
  } catch (e) {
    if (e instanceof SignatureInputParseError) {
      return fail(result, `Signature parse error: ${e.message}`);
    }
    throw e;
  }

  if (
    sigParsed.label &&
    parsedSi.label &&
    sigParsed.label !== parsedSi.label
  ) {
    return fail(
      result,
      `Signature label ${sigParsed.label} does not match Signature-Input label ${parsedSi.label}`,
    );
  }

  // Freshness / replay window (item 1). Runs before the cryptographic check so
  // a stale or expired signature is rejected regardless of whether its bytes
  // would verify. Only covered (signed) created/expires are trusted; see
  // freshness.ts.
  const now = input.now ?? Math.floor(Date.now() / 1000);
  const mode: SigningBaseMode = input.mode ?? "rfc9421";
  try {
    checkFreshness(parsedSi.parameters, parsedSi.covered_components, {
      now,
      maxAgeSeconds: input.maxAgeSeconds,
      maxSkewSeconds: input.maxSkewSeconds,
      enforceExpires: input.enforceExpires,
      requireCreated: input.requireCreated,
      // In rfc9421 mode @signature-params (with created/expires) is signed, so
      // those params are trustworthy even if not covered components.
      paramsSigned: mode === "rfc9421",
    });
  } catch (e) {
    if (e instanceof FreshnessError) {
      return fail(result, `Freshness check failed: ${e.message}`);
    }
    throw e;
  }

  // Anti cross-protocol reuse via the RFC 9421 `tag` parameter (item 2). `tag`
  // rides inside @signature-params, which is signed verbatim, so a present tag
  // is already cryptographically bound; here we enforce the caller's policy.
  const tagValue = parsedSi.parameters["tag"];
  if (input.requireTag && tagValue === undefined) {
    return fail(result, "Signature 'tag' parameter required but absent");
  }
  if (
    input.expectedTag !== undefined &&
    input.expectedTag !== null &&
    tagValue !== input.expectedTag
  ) {
    return fail(
      result,
      `Signature 'tag' ${JSON.stringify(tagValue)} does not match expected ${JSON.stringify(input.expectedTag)}`,
    );
  }

  const requireCd = input.requireContentDigest ?? true;
  // string = enforce that specific algorithm; null or undefined = accept any
  // supported algorithm present in the Content-Digest header (SHA-256 or
  // SHA-512 in this v0.2.x). Default is permissive.
  const requireAlg: string | undefined =
    input.requireAlgorithm == null ? undefined : input.requireAlgorithm;

  if (requireCd) {
    const cdHeader = normHeaders["content-digest"];
    if (!cdHeader)
      return fail(result, "Content-Digest header required but missing");
    // RC-2 body integrity: content-digest must be a COVERED component. Otherwise
    // an attacker swaps the body + recomputes a matching CD header and the
    // signature (which never covered CD) still verifies. Matching the CD to the
    // body is necessary but not sufficient; it must also be signed. (Parity with
    // the Python verifier; caught by rfc9421_hardening_v1 content_digest_not_covered.)
    const coveredLower = parsedSi.covered_components.map((c) =>
      c.trim().replace(/^"+|"+$/g, "").toLowerCase(),
    );
    if (!coveredLower.includes("content-digest")) {
      return fail(
        result,
        "Content-Digest required but not a covered signature component",
      );
    }
    try {
      verifyContentDigest(input.body, cdHeader, requireAlg);
      result.content_digest_valid = true;
    } catch (e) {
      if (e instanceof ContentDigestError) {
        return fail(result, `Content-Digest verification failed: ${e.message}`);
      }
      throw e;
    }
  } else {
    result.content_digest_valid = true;
  }

  let signingBase: string;
  try {
    signingBase = buildSigningBase({
      coveredComponents: parsedSi.covered_components,
      method: input.method,
      authority: input.authority,
      path: input.path,
      scheme: input.scheme ?? "https",
      headers: normHeaders,
      parameters: parsedSi.parameters,
      mode,
      signatureParamsRaw:
        mode === "rfc9421" ? parsedSi.params_block : undefined,
    });
  } catch (e) {
    if (e instanceof SigningBaseError) {
      return fail(result, `Signing-base build error: ${e.message}`);
    }
    throw e;
  }
  result.signing_base = signingBase;

  // Algorithm-downgrade hardening (item 3). Do NOT default a missing alg to
  // ed25519: an absent alg is ambiguous and a downgrade-by-omission vector, so
  // reject it. Then pin the alg to the caller's allow-list before verifying.
  const alg = parsedSi.parameters["alg"];
  if (alg === undefined) {
    return fail(result, "Signature-Input missing 'alg' parameter");
  }
  if (typeof alg !== "string") {
    return fail(
      result,
      `Signature-Input 'alg' parameter must be string, got ${typeof alg}`,
    );
  }
  const allowedAlgorithms = input.allowedAlgorithms ?? ["ed25519"];
  const allowedLower = new Set<string>();
  for (const a of allowedAlgorithms) {
    allowedLower.add(String(a).toLowerCase());
  }
  if (!allowedLower.has(alg.toLowerCase())) {
    const sorted = Array.from(allowedLower).sort();
    return fail(
      result,
      `Signature algorithm ${JSON.stringify(alg)} is not in the allowed set [${sorted.map((s) => JSON.stringify(s)).join(", ")}]`,
    );
  }

  let sigOk: boolean;
  try {
    sigOk = await verifySignature(
      signingBase,
      sigParsed.signature,
      input.publicKey,
      alg,
    );
  } catch (e) {
    if (e instanceof VerifyError) {
      return fail(result, `Signature verification setup error: ${e.message}`);
    }
    throw e;
  }

  if (!sigOk) {
    return fail(result, "Ed25519 signature does not verify against signing base");
  }
  result.signature_valid = true;

  // Single-use nonce / replay check (item 1). Deferred until after the
  // signature verifies so the caller's store is only ever probed with an
  // authentic message, not attacker-supplied junk. This library stays
  // stateless and only asks "has this (nonce, keyid) been seen?".
  if (input.nonceSeen !== undefined) {
    const nonceValue = parsedSi.parameters["nonce"];
    if (nonceValue !== undefined) {
      const keyidValue = String(parsedSi.parameters["keyid"] ?? "");
      let already: boolean;
      try {
        already = input.nonceSeen(String(nonceValue), keyidValue);
      } catch (e) {
        // A store failure must fail closed, not open.
        return fail(
          result,
          `Nonce store check errored: ${(e as Error).message}`,
        );
      }
      if (already) {
        return fail(result, "Signature nonce has already been used (replay)");
      }
    }
  }

  result.valid = result.signature_valid && result.content_digest_valid;
  return result;
}

File: python/algovoi_rfc9421_verifier/verify.py
"""
RFC 9421 + RFC 9530 verification top-level.

Two entry points:

- verify_signature(signing_base, signature_bytes, public_key) ->
  VerifyResult.  Lower-level: caller has already built the signing
  base.
- verify_request(method, authority, path, headers, body, public_key)
  -> VerifyResult.  High-level: parses Signature-Input + Signature,
  builds the signing base, verifies the signature and the
  Content-Digest in one call.

This v0.2.0 surface supports Ed25519 only (the substrate's reference
algorithm; aligned with the RFC 8032 Section 7.1 deterministic
keypair used in the algovoi-jcs-conformance-vectors fixture set).
Other JOSE algorithms (ECDSA-P256, RSA-PSS, HMAC) are roadmap.
"""

from __future__ import annotations

import time
from dataclasses import dataclass, field
from typing import Callable, Union

from algovoi_rfc9421_verifier.parse import (
    parse_signature_input,
    parse_signature_value,
    SignatureInputParseError,
)
from algovoi_rfc9421_verifier.signing_base import (
    build_signing_base,
    SigningBaseError,
)
from algovoi_rfc9421_verifier.content_digest import (
    verify_content_digest,
    ContentDigestError,
)
from algovoi_rfc9421_verifier.freshness import (
    check_freshness,
    FreshnessError,
)


class VerifyError(ValueError):
    """Raised when verification setup is invalid (not when a signature
    fails to verify -- that is captured in VerifyResult.errors)."""


@dataclass
class VerifyResult:
    """Result of an RFC 9421 verification.

    valid is True only if every check ran AND every check passed.
    Specific check outcomes are exposed individually for debugging.
    """

    valid: bool = False
    signature_valid: bool = False
    content_digest_valid: bool = False
    signing_base: str = ""
    covered_components: list[str] = field(default_factory=list)
    parameters: dict = field(default_factory=dict)
    label: str = ""
    errors: list[str] = field(default_factory=list)

    def fail(self, msg: str) -> "VerifyResult":
        self.errors.append(msg)
        self.valid = False
        return self


PublicKey = Union[bytes, str]
"""Either raw Ed25519 public-key bytes (32 bytes) or hex-encoded string."""


def _public_key_bytes(public_key: PublicKey) -> bytes:
    if isinstance(public_key, bytes):
        if len(public_key) != 32:
            raise VerifyError(
                f"Ed25519 public key must be 32 bytes, got {len(public_key)}"
            )
        return public_key
    if isinstance(public_key, str):
        key_hex = public_key.removeprefix("0x")
        try:
            kb = bytes.fromhex(key_hex)
        except ValueError as e:
            raise VerifyError(f"public key hex is invalid: {e}") from e
        if len(kb) != 32:
            raise VerifyError(
                f"Ed25519 public key hex must decode to 32 bytes, got {len(kb)}"
            )
        return kb
    raise VerifyError(
        f"public_key must be bytes or hex str, got {type(public_key).__name__}"
    )


def verify_signature(
    signing_base: str,
    signature_bytes: bytes,
    public_key: PublicKey,
    *,
    algorithm: str = "ed25519",
) -> bool:
    """Verify an Ed25519 signature over the signing base.

    Returns True on success. Raises VerifyError on invalid inputs.
    Returns False on signature mismatch.
    """
    if algorithm.lower() != "ed25519":
        raise VerifyError(
            f"v0.1.0 supports ed25519 only; got {algorithm!r}"
        )
    if not isinstance(signing_base, str):
        raise VerifyError(
            f"signing_base must be str, got {type(signing_base).__name__}"
        )
    if not isinstance(signature_bytes, (bytes, bytearray)):
        raise VerifyError(
            f"signature_bytes must be bytes-like, got {type(signature_bytes).__name__}"
        )
    if len(signature_bytes) != 64:
        raise VerifyError(
            f"Ed25519 signature must be 64 bytes, got {len(signature_bytes)}"
        )

    pk_bytes = _public_key_bytes(public_key)

    try:
        from nacl.signing import VerifyKey
        from nacl.exceptions import BadSignatureError
    except ImportError as e:
        raise VerifyError(
            "PyNaCl is required for Ed25519 verification. Install with 'pip install pynacl'."
        ) from e

    verify_key = VerifyKey(pk_bytes)
    try:
        verify_key.verify(
            signing_base.encode("utf-8"), bytes(signature_bytes)
        )
    except BadSignatureError:
        return False
    return True


def verify_request(
    *,
    method: str,
    authority: str,
    path: str,
    headers: dict[str, str],
    body: bytes,
    public_key: PublicKey,
    scheme: str = "https",
    require_content_digest: bool = True,
    require_algorithm: str | None = None,
    mode: str = "rfc9421",
    now: int | None = None,
    max_age_seconds: int | None = None,
    max_skew_seconds: int = 60,
    enforce_expires: bool = True,
    require_created: bool = False,
    expected_tag: str | None = None,
    require_tag: bool = False,
    allowed_algorithms: frozenset[str] | set[str] = frozenset({"ed25519"}),
    nonce_seen: Callable[[str, str], bool] | None = None,
) -> VerifyResult:
    """High-level verification of an RFC 9421-signed HTTP request.

    Steps:
      1. Locate Signature-Input + Signature headers
      2. Parse Signature-Input -> covered components + parameters
      3. Parse Signature -> signature bytes
      4. (Optional) Verify Content-Digest against the body
      5. Build the signing base from covered components
      6. Verify the Ed25519 signature

    Returns a VerifyResult. All errors land in result.errors;
    result.valid is True iff every check passed.

    mode selects the signing-base construction:
      - "rfc9421" (default): RFC 9421 §2.5 compliant — @method
        case-preserved, @signature-params line appended. Matches output
        of algovoi-rfc9421-signer and any other compliant implementation.
      - "algovoi-v0": legacy internal format — @method lowercased, no
        @signature-params line. Kept for backward compatibility with the
        rfc9421_proxy_chain_v0 conformance fixture set signed before v0.2.0.
    """
    result = VerifyResult()

    norm_headers = {k.lower(): v for k, v in headers.items()}

    si_value = norm_headers.get("signature-input")
    if not si_value:
        return result.fail("Signature-Input header missing")
    s_value = norm_headers.get("signature")
    if not s_value:
        return result.fail("Signature header missing")

    try:
        parsed_si = parse_signature_input(si_value)
    except SignatureInputParseError as e:
        return result.fail(f"Signature-Input parse error: {e}")
    result.label = parsed_si.label
    result.covered_components = parsed_si.covered_components
    result.parameters = parsed_si.parameters

    try:
        s_label, sig_bytes = parse_signature_value(s_value)
    except SignatureInputParseError as e:
        return result.fail(f"Signature parse error: {e}")
    # Enforce label match only when both labels are non-empty. The
    # unlabelled forms found in some real-world implementations don't
    # carry labels at all, so any match against an empty label is
    # treated as compatible.
    if (
        s_label
        and parsed_si.label
        and s_label != parsed_si.label
    ):
        return result.fail(
            f"Signature label {s_label!r} does not match Signature-Input label {parsed_si.label!r}"
        )

    # Freshness / replay window (item 1). Runs before the cryptographic check so a
    # stale or expired signature is rejected regardless of whether its bytes would
    # verify. Only covered (signed) created/expires are trusted; see freshness.py.
    if now is None:
        now = int(time.time())
    try:
        check_freshness(
            parsed_si.parameters,
            parsed_si.covered_components,
            now=now,
            max_age_seconds=max_age_seconds,
            max_skew_seconds=max_skew_seconds,
            enforce_expires=enforce_expires,
            require_created=require_created,
            # In rfc9421 mode @signature-params (with created/expires) is signed,
            # so those params are trustworthy even if not covered components.
            params_signed=(mode == "rfc9421"),
        )
    except FreshnessError as e:
        return result.fail(f"Freshness check failed: {e}")

    # Anti cross-protocol reuse via the RFC 9421 `tag` parameter (item 2). `tag`
    # rides inside @signature-params, which is signed verbatim, so a present tag is
    # already cryptographically bound; here we enforce the caller's policy on it.
    tag_value = parsed_si.parameters.get("tag")
    if require_tag and tag_value is None:
        return result.fail("Signature 'tag' parameter required but absent")
    if expected_tag is not None and tag_value != expected_tag:
        return result.fail(
            f"Signature 'tag' {tag_value!r} does not match expected {expected_tag!r}"
        )

    if require_content_digest:
        cd_header = norm_headers.get("content-digest")
        if not cd_header:
            return result.fail("Content-Digest header required but missing")
        # D-F2: body integrity requires content-digest to be a COVERED component.
        # Otherwise an attacker swaps the body + recomputes a matching CD header and
        # the signature (which never covered CD) still verifies. Matching the CD to
        # the body is necessary but not sufficient -- it must also be signed.
        _covered = [c.lower().strip().strip('"') for c in parsed_si.covered_components]
        if "content-digest" not in _covered:
            return result.fail("Content-Digest required but not a covered signature component")
        try:
            verify_content_digest(
                body, cd_header, require_algorithm=require_algorithm
            )
            result.content_digest_valid = True
        except ContentDigestError as e:
            return result.fail(f"Content-Digest verification failed: {e}")
    else:
        result.content_digest_valid = True

    try:
        signing_base = build_signing_base(
            parsed_si.covered_components,
            method=method,
            authority=authority,
            path=path,
            scheme=scheme,
            headers=norm_headers,
            parameters=parsed_si.parameters,
            mode=mode,
            signature_params_raw=parsed_si.params_block if mode == "rfc9421" else None,
        )
    except SigningBaseError as e:
        return result.fail(f"Signing-base build error: {e}")
    result.signing_base = signing_base

    # Algorithm-downgrade hardening (item 3). Do NOT default a missing alg to
    # ed25519: an absent alg is ambiguous and a downgrade-by-omission vector, so
    # reject it. Then pin the alg to the caller's allow-list before verifying.
    alg = parsed_si.parameters.get("alg")
    if alg is None:
        return result.fail("Signature-Input missing 'alg' parameter")
    if not isinstance(alg, str):
        return result.fail(
            f"Signature-Input 'alg' parameter must be string, got {type(alg).__name__}"
        )
    if alg.lower() not in {a.lower() for a in allowed_algorithms}:
        return result.fail(
            f"Signature algorithm {alg!r} is not in the allowed set "
            f"{sorted(allowed_algorithms)}"
        )

    try:
        sig_ok = verify_signature(
            signing_base, sig_bytes, public_key, algorithm=alg
        )
    except VerifyError as e:
        return result.fail(f"Signature verification setup error: {e}")
    if not sig_ok:
        return result.fail("Ed25519 signature does not verify against signing base")
    result.signature_valid = True

    # Single-use nonce / replay check (item 1). Deferred until after the signature
    # verifies so the caller's store is only ever probed with an authentic message,
    # not with attacker-supplied junk. The caller records the nonce on success; this
    # library stays stateless and only asks "has this (nonce, keyid) been seen?".
    if nonce_seen is not None:
        nonce_value = parsed_si.parameters.get("nonce")
        if nonce_value is not None:
            keyid_value = str(parsed_si.parameters.get("keyid", ""))
            try:
                already = nonce_seen(str(nonce_value), keyid_value)
            except Exception as e:  # a store failure must fail closed, not open
                return result.fail(f"Nonce store check errored: {e}")
            if already:
                return result.fail("Signature nonce has already been used (replay)")

    result.valid = result.signature_valid and result.content_digest_valid
    return result
